splice_machine/src/main/java/com/splicemachine/derby/impl/sql/execute/operations/DMLWriteOperation.java

```diff
@@ -179,15 +182,44 @@ protected TxnView getTransactionForWrite(DataSetProcessor dsp) throws StandardEx
         try {
             if (dsp.getType() == DataSetProcessor.Type.SPARK || isOlapServer())
             {
-                nestedTxn =
+                WriteCursorConstantOperation constantAction=(WriteCursorConstantOperation)writeInfo.getConstantAction();
+                TriggerInfo triggerInfo=constantAction.getTriggerInfo();
+                boolean needsfullyNestedTransaction = triggerInfo != null;
+
+                LanguageConnectionContext lcc = getActivation().getLanguageConnectionContext();
+                TransactionController transactionExecute = lcc.getTransactionExecute();
+                if (needsfullyNestedTransaction && !lcc.hasNestedTransaction()) {
+                    // Start a nested transaction controller that deals only with internal
+                    // transactions, not savepoints.
+                    transactionController =
+                        transactionExecute.startNestedInternalTransaction(false,
+                                            Bytes.toBytes(Long.toString(heapConglom)),
+                                            false);
+                    lcc.pushNestedTransaction(transactionController);
+                    Transaction rawStoreXact = ((TransactionManager) transactionController).getRawStoreXact();
+                    if (!(((BaseSpliceTransaction) rawStoreXact).getActiveStateTxn() instanceof Txn))
+                        throw StandardException.newException(LANG_INTERNAL_ERROR,
+                                        "DMLWriteOperation cannot make child Txn out of TxnView.");
+                    Txn childTxn = (Txn) ((BaseSpliceTransaction) rawStoreXact).getActiveStateTxn();
+                    nestedTxn = childTxn;
+                    txn = (TxnView) childTxn;
+                }
+                else {
+                    nestedTxn =
                     SIDriver.driver().lifecycleManager().beginChildTransaction(
-                            parent,
-                            parent.getIsolationLevel(),
-                            parent.isAdditive(),
```

Comment:
I don't quite understand this branch. It seems right in case of `!needsfullyNestedTransaction`. But if it's `needsfullyNestedTransaction && lcc.hasNestedTransaction()`, I assume that's because some other write operation enters the branch above earlier? In that case, what does the code mean in this branch? Does it start a child transaction in a different way and push into a different stack? But what happens if a third write operation comes?

Sorry this might be a stupid question. I haven't worked on transactions in Splice yet.

Reply:
@ascend1 Not a stupid question at all.  The first time around, lcc.pushNestedTransaction needs to be called to create a SpliceInternalTransactionManager instead of the default transaction manager.  This special transaction manager has the ability to do full nesting of transactions.  The first time around, we don't need to create a nested transaction, we are just using the main transaction that was elevated to writable on line 85 of QueryJob.java:
    TxnView parent = root.getCurrentTransaction();

Every time after this that we create a new child transaction (because we have a new nested operation) and this transacction is pushed onto the transaction manager's transaction stack at line 220:

internalTransaction.pushInternalTransaction((Txn) txn); 

So, we can go to the else branch once the transaction manager that supports fully nested transactions is set up, and we only create the child transaction on the 2nd, 3rd, 4th, etc. nested operations.

Maybe the method name hasNestedTransaction is causing the confusion.  Do you think I should change the name to something like hasSpliceInternalTransactionManager() ?